server/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -697,6 +697,6 @@ public long getAutoGeneratedTimestamp() {
 
     @Override
     public long ramBytesUsed() {
-        return SHALLOW_SIZE + RamUsageEstimator.sizeOf(id) + (source == null ? 0 : source.ramBytesUsed());
+        return SHALLOW_SIZE + RamUsageEstimator.sizeOf(id) + (source == null ? 0 : source.length());
```

Comment:
While I think that this should be ok for now, it's critical to confirm though that our accounting is not completely off with some tests/benchmarks.

A first test to ensure that we are not vastly underaccounting should have a high `indices.write.limit` (e.g. 80%) and check that we don't hit the real-memory circuit breaker when we are filling up the node (should probably block requests from being processed), but rather the `indices.write.limit` instead.
A second test to ensure that we are not vastly overaccounting should have a low `indices.write.limit`, for example (e.g. 100MB), and show that we can still put a decently large number of requests to the system (i.e. send bulks that are in total making up 80MB in input size) without reaching the limit.
Third we need to check what the overhead of primaries is to wait on replica responses (i.e. memory consumed by listeners etc.). For that we should block processing on a replica, and fill up a primary and see how much memory is being consumed.